R/shell_connection.R

```diff
@@ -546,34 +546,82 @@ initialize_connection.spark_shell_connection <- function(sc) {
       }
 
       context_config <- connection_config(sc, "spark.", c("spark.sql."))
-      apply_config(context_config, conf, "set", "spark.")
+      apply_config(conf, context_config, "set", "spark.")
 
       default_config <- shell_connection_config_defaults()
       default_config_remove <- Filter(function(e) e %in% names(context_config), names(default_config))
       default_config[default_config_remove] <- NULL
-      apply_config(default_config, conf, "set", "spark.")
+      apply_config(conf, default_config, "set", "spark.")
 
       # create the spark context and assign the connection to it
-      sc$spark_context <- invoke_static(
-        sc,
-        "org.apache.spark.SparkContext",
-        "getOrCreate",
-        conf
-      )
+
+      sc$spark_context <- if (spark_version(sc) >= "2.0") {
+
+        # For Spark 2.0+, we create a `SparkSession`.
+        session <- invoke_static(
+          sc,
+          "org.apache.spark.sql.SparkSession",
+          "builder"
+        ) %>%
+          invoke("config", conf) %>%
+          apply_config(connection_config(sc, "spark.sql."), "config", "spark.sql.") %>%
+          invoke("getOrCreate")
+
+        # Cache the session as the "hive context".
+        sc$state$hive_context <- session
+
+        # Return the `SparkContext`.
+        invoke(session, "sparkContext")
+      } else {
+        ctx <- invoke_static(
+          sc,
+          "org.apache.spark.SparkContext",
+          "getOrCreate",
+          conf
+        )
+
+        sc$state$hive_context <- tryCatch(
+          invoke_new(sc, "org.apache.spark.sql.hive.HiveContext", ctx),
+          error = function(e) {
+            warning(e$message)
+            warning("Failed to create Hive context, falling back to SQL. Some operations, ",
+                    "like window-functions, will not work")
+
+            jsc <- invoke_static(
+              sc,
+              "org.apache.spark.api.java.JavaSparkContext",
+              "fromSparkContext",
+              ctx
+            )
+
+            hive_context <- invoke_static(
+              sc,
+              "org.apache.spark.sql.api.r.SQLUtils",
+              "createSQLContext",
+              jsc
+            )
+
+            params <- connection_config(sc, "spark.sql.")
+            apply_config(hive_context, params, "setConf", "spark.sql.")
+
+            # return hive_context
+            hive_context
+          }
+        )
+
+        ctx
+      }
 
       invoke(backend, "setSparkContext", sc$spark_context)
     }
 
-    sc$spark_context$connection <- sc
```

Comment:
We probably still need this one and the `java_context` bellow.

When we call `invoke_static()` with an `sc`, the `sc` becomes attached to the object, this is to be able to execute `invoke()` on the object etc. This is fine for most operations; however, if you need to get the `spark_context()` and then indirectly get the `spark_context()` from there again, you would get a `spark_context()` that is no longer there. The way we patched this was by reattaching the connection once the `spark_context()` is created.